### DISCUSSION TITLE
How to use the class SecureArray?

### DISCUSSION THREAD

**Discussion by 1191000814** [lschoe/mpyc#81] [Q&A] [answered]
How to use the class SecureArray? It is not seem in the document? I want to transfer a np.ndarray type to secure type, how can I do that?

**Comment by lschoe** [comment]
It's best to look at a demo like [demos/np_id3gini.py](https://github.com/lschoe/mpyc/blob/90165a8f3e6c00792ab68ab310287ccec2cefd3a/demos/np_id3gini.py) to see how to work with secure arrays. Also see the other demos listed in [demos/np-run-all.sh](https://github.com/lschoe/mpyc/blob/90165a8f3e6c00792ab68ab310287ccec2cefd3a/demos/np-run-all.sh.)

For a basic secure type like `secint = mpc.SecInt()` you can use `b = secint.array(a)` to convert a Numpy array `a` into a secret-shared version `b`, and then work with `b` using Numpy operations.

**Comment by 1191000814** [comment]
Thanks! I have another problem, the SecInt type is unhashable, but I want to use it as the key of a dict, How can I do?

> It's best to look at a demo like [demos/np_id3gini.py](https://github.com/lschoe/mpyc/blob/90165a8f3e6c00792ab68ab310287ccec2cefd3a/demos/np_id3gini.py) to see how to work with secure arrays. Also see the other demos listed in [demos/np-run-all.sh](https://github.com/lschoe/mpyc/blob/90165a8f3e6c00792ab68ab310287ccec2cefd3a/demos/np-run-all.sh.)
> 
> For a basic secure type like `secint = mpc.SecInt()` you can use `b = secint.array(a)` to convert a Numpy array `a` into a secret-shared version `b`, and then work with `b` using Numpy operations.

**Comment by lschoe** [comment, marked as answer]
Instances of secure types like secure integers are not hashable because their values are secret-shared. Each party holds a share of the value, but these shares are random and different for each party.

Depending on what you need to do, you might use operators like `==` and `<` with secure integers to construct searchable datastructures but this may get quite involved.

**Reply by 1191000814** [reply]
Hello, I have another problem. when I use mpc.output() return **a list which contains some lists**, there is a error. It mean "AttributeError: 'list' object has no attribute 'field' ". So How can I return the data strcture (a list which contains some lists)

**Reply by lschoe** [reply]
In that case you can use a list comprehension iterating over nested lists. In your case of a list of lists (which need not be of the same length) `s`, you can do something like `[await mpc.output(_) for _ in s]`.

**Reply by 1191000814** [reply]
Hello, for per mpc.pid, if I want to `mpc.input`  **a list which contains some lists**, How should I do? It show that `AttributeError: 'list' object has no attribute 'shape'`

**Reply by 1191000814** [reply]
Except for mpc.if_else(), is there any other methods to implement if-else clause?

**Reply by 1191000814** [reply]
like this， I want to return [1, 2] if the condition is established，and return [1, 2, 3] else.
`async def main():
    async with mpc:
        a = sec_int(1)
        ic(a)
        a = mpc.input(a)[0]
        # a = mpc.np_fromlist(a[0])  # .reshape([3, ])
        b = mpc.if_else(a >= sec_int(2), [1, 2], [1, 3, 4])
        ic(b)
        # b = await mpc.output(b)
        # ic(b)


if __name__ == '__main__':
    mpc.run(main())
`
but it return [1, 3, 4] always. Why? (and a error is comming soon) 
Thanks a lot for you help!

**Reply by lschoe** [reply]
To work with `mpc.input()` for nested lists, use a similar approach as suggested for `mpc.output()` above. For instance, use `[mpc.input(_, senders=0) for _ in s]` if you want only party 0 to provide private input, and `s` is a list of lists of secure integers, say. Drop the `senders=0` argument if you want all parties to have private inputs. 

For `mpc.if_else()`,  note that the lists for both branches should be of the same length. You cannot obliviously select between lists of different lengths like you're doing with `[1,2]` and `[1,3,4]` because the length of the result would already reveal which branch was followed. Also, make sure to cast the elements of these lists to secure integers, So something like this should work:

```python
b = mpc.if_else(a >= secint(2), [secint(1), secint(2), secint(0)], [secint(1), secint(3), secint(4)])
```

If you'd work with secure Numpy arrays, there you also have `np.where()` that can be used with secure integer arrays, as you can see in these test cases for example: https://github.com/lschoe/mpyc/blob/a11ff6a859a049898185db5ce1363ec121318337/tests/test_runtime.py#L120-L123

**Reply by 1191000814** [reply]
Thanks for your patience!
For instance, if I want to add the indexes(not unique) of list `my_list` which value is `a` to another list `b` , How should I do?

**Reply by lschoe** [reply]
There are several ways to do this. Maybe something like this suits your purposes:
```python
t = [a == _ for _ in my_list]
u = mpc.seclist([-1] * (len(s)+1), secint)
for i in range(len(s), 0, -1):
    u[mpc.sum(t[:i])] = i-1
b = list(u[1:])
```

**Reply by 1191000814** [reply]
Thanks for you reply!

Maybe there is still a problem, the seclist `u` contains some other elements because it should ensure its length in the first,  so that I will have many troubles in solving the tail elements(I do not  know the number of elements, and the seclist can not be **range indexed** by secint). I want the list `u` contain no redundant elements.

But... I observed the class `secindex`, will it be a good idea for the problem? I can not understand anything about `secindex` until now because of the concise documentation.

**Reply by lschoe** [reply]
I wouldn't go there. Actually, if you do not need to hide the number of occurrences of `a `in `my_list`, then you can first reveal this number, using `n = await mpc.output(mpc.sum(t))`, and then use `n` instead of `len(s)`in what follows. Like this:
```python
t = [a == _ for _ in my_list]
n = await mpc.output(mpc.sum(t))
u = mpc.seclist([-1] * (n+1), secint)
for i in range(len(my_list), 0, -1):
    u[mpc.sum(t[:i])] = i-1
b = list(u[1:])
await mpc.output(b)
```

**Comment by 1191000814** [comment]
Hi, today when I use **await mpc.output()** to get the real data of a variable, **the program was stuck** but not end or error, I don't know why it happens. Is there a dead lock or other reason?

**Reply by 1191000814** [reply]
I used `await mpc.out()` many times in a program, and it got stuck at one point. It was `n = await mpc.output(secure_updated.count(-1))`, but if I used a simple clause like `n = await mpc.output( secint(3))`, it is normal; I'm sorry to bother you, but this problem has puzzled me for a long time.

**Reply by 1191000814** [reply]
By the way, I use -1 as a padding num to make muti-lists the same length before, when I use -2 as the padding num, it can run more long time, when I use 100, it become normal, it' s so strangeness.

All the way, Thanks for your help!
